[Merged by Bors] - currency docs (#1305)
ComposableFi/composable: docs/currency.md

# Overview

This documents describes various aspects of currency. All these aspects are in some form encoded into protocol.

Document uses next terms interminably:

- `currency`, `asset`
- `this parachain`, `consensus`

## What is currency?

> Identity of currency is positive integer. Any positive integer may be currency id, but not all integers are currency ids.

Reviewer: ```suggestion
The identity of a currency is a positive integer. Any positive integer may be currency id, but not all integers are currency ids.
```

All currencies created in runtime are managed by [CurrencyFactory](../frame/currency-factory) pallet. Look into pallet readme for clarification of low level basics.

### Amounts

Each currency has a non negative total issuance. Given id, we can ask for total supply of currency in consensus of ledger(this parachain).  Amount of 10 of currency is equal to amount of 10 of same currency - it is fungible.

### Decimals

Currency may have unit and minimal amount. Unit usually 10 to some power like 6, 9 or `12`. Unit usually is priceable and comprehendible by people.  While minimal amount can be used to operate micro transactions and help to minimize rounding errors.

Most pallets operate in decimals oblivion manner. Only pallets which care about out of consensus metadata operate with decimals. Examples are oracle and bridges protocols.

Ledgers operate only in decimal currencies. Zero decimals is viable currency too.

## Remote or local

All currencies are eventually local and have local identifiers.

> Any remote currency bridged to our parachain is minted in equal amount what is bridged.

Author: amount

Remote currency minted and burn equal amount it was transferer into and out of parachain.

> Remote token on our chain can be local on other chain or be remote too.

Author: idea that can have one chain get remote token and get that token to us. and we will think it is that chain token, when really it is other chain token. original. so chain acts as bridge.

Semantically in local consensus bridged tokens are just protocol tokens with their associate risks.

Check [AssetsRegistry](../frame/assets-registry) pallet for details of mapping.

### Remote currencies equivalence

In general, remote currency transferred via one bridge is not same, in general, as currency transferred via other bridge.

Example, USDC transferred via Mosaic is not equal to amount transferred via XCMP from Acala.

There are several solution to make currencies equal:

- create DEX for pair and user router to swap. if currencies are equal in price, swap would be 1 to 1.
- embed direct trust into bridges, but configuring setting ratio  of currencies to be 1 to 1. this is risky.
- use trustless bridging, in this there are proves of equivalence. risk are low.

> XCMP allows low risk form of transfer by using `Reserve transfer` (any transfer goes via chain which issued token). So XCMP allows `Teleport` which requires having trust among several parachains.

Author: also

IBC MMR tokens are of lowest risk.

Mosaic has risks of centralized bridges.

### Remote currency identifiers

Any remote currency has bidirectional map from and to local.
That allows send tokens amount from and to local consensus.

Remote currencies usually have related bridge kind/network(chain) identifiers attached.

## Native currencies and gas fees

Users pay on fees for execution in native currency for runtime.

> There is runtime configuration which allows to map amount of execution which can be bought for amount of native currency.

Author: will rewrite

This parachain uses pre calculated weights which maps to hardware units. Amount of weight for native currency which can be consumed for native currency is non linear function.

Cross chain interactions over bridges may involve metered networks which allow setting limit on execution price and calculate amount of executed in runtime.

> There are non paying `native` currencies used inside of protocols built on top of runtime. These are used for governance's and revenue distribution, but not for paying for execution.

Author: will rewrite

### Bridged gas fee (`bring your own gas`)

When user whats to have assets on our chain, he either can be minted with some native currency, or it can be allowed to pay for cross chain transaction in some  on native currency.

Currently there non native currency is mapped to native with some ratio which then mapped to weight.

> Likely the transaction issues by user should swap some remote currencies into native on transfer to proceed to execute local transactions.

Author: will rewrite

## Tokenomics

So any currency is eventually local and minting is governed by `protocol`.

Simplest protocol is initial minting by `root` or `governance voting`.

More complicated are `bridges` and `staking` or `LP`.

> 

Author: bridge is protocol and staking. will rewrite

Currently there non native currency is mapped to native with some ratio which then mapped to weight.

Likely the transaction issues by user should swap some remote currencies into native on transfer to proceed to execute local transactions.

## Tokenomics

So any currency is eventually local and minting is governed by `protocol`.

Simplest protocol is initial minting by `root` or `governance voting`.

More complicated are `bridges` and `staking` or `LP`.

### Inflation

Less other assets can be obtained for same amount of inflated asset than previous.

Inflation may happen when people exchange inflating currency into others currencies a lot.

May happen when more tokens are minted for inflating currency. Releasing locked currency to the market make inflate it too.

Shorting some currency or off chain issues (bridges bugs or offchain oracle information) may lead to inflation too.

There are protocols, which try to keep mint tokens to maintain some level price and liquidity.
Examples are stable coin and staking protocols.

> Provably burning/locking tokens usually leads to deflation.

Author: misleading offchain oracle

### Dilution

Some protocols allows users to put tokens into and issues some tokens which represent total share of protocol.

Protocol may issues more tokens, may be when new users get into.

That leads to reduction of total user share in protocol, so his tokens are diluted.

### Risk categories

Some assets may be considered more volatile/riskier than others.

That consideration is not constrained by protocol, but brings important, oracle like, information of chain.

> Used in protocols, like lending or credit scores.

Author: will rewrite

# Overview

This document describes various aspects of currency as they relate to composable finance. All these aspects are, in some form, embedded into our protocols.

In this document, we use the following terms interchangeably:

- `currency`, `asset`, `token`
- `this parachain`, `consensus`

## What is currency?

> We express the identity of a currency as a positive integer. Any positive integer may be a currency id, but not all integers are currency ids.

Author: https://www.google.com/search?q=positive+number&oq=positive+number&aqs=chrome..69i57.1914j0j1&sourceid=chrome&ie=UTF-8

so positive is by definion > 0

The [CurrencyFactory](../frame/currency-factory) pallet manages all currencies created in the runtime. For clarification on the low-level basics, look into the pallet readme.

### Amounts

Each currency has a non-negative total issuance. Given an identifier, we can ask for the total supply of the respective currency in consensus with the ledger(in this case, the parachain). A Currency is considered fungible because an amount of 10 of said currency equals another 10 of the same currency on another account.

### Decimals

A Currency may be associated with a `unit` and `minimal amount`. `Unit` usually comes in an amount of 10 to some power like 6, 9, or 12. The units are typically priceable and comprehendible for users. `minimal amount` can be used to operate micro-transactions and help minimize rounding errors.  

> Most pallets operate in a decimals oblivion manner. Meaning these pallets execute all mathematical operations on whole numbers.

Reviewer 2: ```suggestion
Most pallets operate with no knowledge of decimals and  execute all mathematical operations on whole numbers.
```

> For example, currency A has 12 decimals. 420 units of A would be represented by the operation 420*10^12. Currency's B 420 units could be 420*10^09. The amount of A would be 42*10^13 as used in pallets' math, while the amount of B would be 42*10^16. Pallets use larger numbers internally to keep errors out of mathematical operations, rounding effectively non on existent per protocol basis. 

Author: oh, will fix. will make less misleading.

For example, ETH is remote on Acala and AUSD is local on Acala. Both are remote on this parachain and we consider them be from Acala both. 

In the local consensus, bridged tokens semantically are just protocol tokens with their associated risks.

For details on registry mapping, check the [AssetsRegistry](../frame/assets-registry) pallet. 

### Remote currencies equivalence

Generally, a remote currency transferred via one bridge is not necessarily the same as the same currency transferred via another bridge.

For example, USDC transferred via Mosaic is not equal to the same amount transferred via XCMP from Acala.

There are several solutions to equalize currencies:

- Create a DEX for the currency pair and use a router to swap them. For example, if the currencies are equal in price, the swap would be 1 to 1. Market defined is best for high-risk bridges. 
- A risky approach is to embed direct trust to bridges via the configured ratio of transferred currencies to be 1 to 1.
- Use trustless decentralized bridging. With this solution, there is proof of equivalence, making it the lowest risk. For example, IBC MMR or XCMP `Reserver Transfer`.

> For details, read XCVM documentation of security and risks of choosing bridges(relayers).

Author: SPEC.md has some, but not sure if it is best. so only when separate bridges.md will be likely it makes sence to link.

For example, ETH was transferred from Etherium to Acala and to this parachain as AETH. Same time METH got via Moonbeam. Moonbeam was hacked, and minted a huge amount of unbacked METH hacker wants. In this the case DEX(market) would react and skew the price of AETH to be higher than METH. 
More mild form of non equivalence would be becasue of network congestion. Imagine Acala transfer prices are several time more of Moonbeam. In this case  AETH will be less liquid than METH, so it price may be lower. 

### Remote currency identifiers

Any remote currency has a bidirectional map from and to local.
Bidirectional mapping allows sending the number of tokens from and to local consensus.

Remote currencies usually have their respective bridge identifiers attached in the format "kind/network(chain)".

## Native currencies and gas fees

Users pay fees in the runtimes native currency for the execution. It allows the runtime to earn and prevents spam and DDoS.

> A runtime configuration allows mapping amount of execution which can be bought for amount of native currency.

Author: configuration in our transaction fees pallet hook. i think links can be improve later. so I guess payments will be separate document later.

This parachain uses pre-calculated weights to estimate hardware costs of execution. An amount of weight a native currency can buy is a non-linear dynamic function from currency amount to weight. As the usage of this chain goes to maximum capacity, a unit of weight costs more and more. 

Cross-chain interactions over bridges may involve metered networks which allow setting limits on execution price and calculating the possible number of execution resources in the runtime. For example, Ethereum.

There are non-paying `native` currencies used inside of protocols built on top of the runtime. They are used for governance and revenue distribution but do not pay for execution.

> ### Bridged gas fee (`bring your own gas`)

Author: it is abbreviation, will leave it quoted kind of

> Users who want to have assets on our chain can either be minted with some native currency or be allowed to pay for cross-chain transactions in some non-native currency.

Author: minted by governance

Currently, non-native currencies are mapped to native with some ratio which in turn is mapped to a weight.

Transactions issued by users are likely to swap some remote currencies into native on transfer, to enact the execution of local transactions. 

## Tokenomics

So any currency is eventually local, and minting is governed by a protocol.

The most straightforward protocol is initial minting by `root` or `governance voting`.

Other more complicated protocols are `bridges`, `staking`, or `LP` are more complicated.

> Protocols form various economic facets of tokens.

Author: commented

### Inflation

In the context of currency, we define inflation as when the amount of an asset can obtain fewer other assets than the same amount of the inflated asset could before.

> Inflation may happen when: 

Reviewer 2: ```suggestion
Inflation may occur when: 
```

> - Users exchange an inflating currency into other currencies a lot.

Reviewer 2: ```suggestion
- Users frequently exchange an inflating currency into other currencies.
```

- An excessive amount of tokens are minted to inflate a currency. 

> - Previously locked currency is released to the market.

Reviewer 2: ```suggestion
- A previously locked currency is released to the market.
```

Reviewer 2: This is referring to the inflated asset having a locked amount released, correct?

- A currency is being shorted 
- Off-chain issues like bridging bugs or misleading off-chain oracle information

> Some protocols try to maintain a specific price level or level of liquidity by continuously minting tokens.
> 
> Examples are stable coins and staking protocols.

Reviewer 2: ```suggestion
Some protocols, such as stable coins and staking, try to maintain a specific price level or level of liquidity by continuously minting tokens.
```

> Burning and locking tokens is probable to result in deflation generally.

Reviewer 2: ```suggestion
Burning and locking tokens will in general cause deflation, as the total available supply has been decreased.
```

### Dilution

Some protocols allow users to provide their tokens to the protocol to issue other tokens, representing the total share of the protocol.
Meaning new users coming into the protocol and issuing more tokens leads to a reduction of the total user share in the protocol diluting their tokens.

### Risk categories

Some assets may be considered more volatile/riskier than others.

> Categories represent governance and oracle-like information on the chain.

Reviewer 2: What is meant by categories here?

In this document, we use the following terms interchangeably:

- `currency`, `asset`, `token`
- `this parachain`, `consensus`

## What is currency?

We express the identity of a currency as a positive integer. Any positive integer may be a currency id, but not all integers are currency ids.

The [CurrencyFactory](../frame/currency-factory) pallet manages all currencies created in the runtime. For clarification on the low-level basics, look into the pallet readme.

### Amounts

Each currency has a non-negative total issuance. Given an identifier, we can ask for the total supply of the respective currency in consensus with the ledger(in this case, the parachain). A Currency is considered fungible because an amount of 10 of said currency equals another 10 of the same currency on another account.

### Decimals

A Currency may be associated with a `unit` and `minimal amount`. `Unit` usually comes in an amount of 10 to some power like 6, 9, or 12. The units are typically priceable and comprehendible for users. `minimal amount` can be used to operate micro-transactions and help minimize rounding errors.  

> Most pallets operate with no knowledge of decimals and  execute all mathematical operations on whole numbers.

Reviewer 3: ```suggestion
Most pallets operate without knowledge of decimals and execute all mathematical operations on whole numbers.
```

Author: will fix in next pr
